File: kernel/daterange.cpp
#include "daterange.h"

DateRange::DateRange()
{
}

DateRange::DateRange(QDate const& start, QDate const& end) :
   m_interval(start, end),
   m_startDate(start)
{
}

DateRange::DateRange(QDate const& start, Interval const& interval) :
   m_interval(interval),
   m_startDate(start)
{
}

uint DateRange::days() const
{
   return m_startDate.daysTo(m_startDate + m_interval);
}

QDate DateRange::endDate() const
{
   return (m_startDate + m_interval).addDays(-1);
}

DateRange DateRange::intersect(DateRange const& other) const
{
   if (isNull() || other.isNull())
   {
      return DateRange();
   }

   QDate startDate = std::max(m_startDate, other.startDate());
   QDate endDate = std::min(this->endDate(), other.endDate());
   if (startDate <= endDate)
   {
      return DateRange(startDate, endDate);
   }
   else
   {
      return DateRange();
   }
}

bool DateRange::isNull() const
{
   return m_startDate.isNull();
}

QDate DateRange::startDate() const
{
   return m_startDate;
}

DateRange& DateRange::operator++()
{
   m_startDate = m_startDate + m_interval;
   return *this;
}

bool DateRange::operator==(DateRange const& other) const
{
   return m_startDate == other.m_startDate && endDate() == other.endDate();
}

File: src/main/java/cashpiles/time/DateRange.java
package cashpiles.time;

import java.math.BigInteger;
import java.time.DateTimeException;
import java.time.LocalDate;
import java.time.Period;
import java.time.temporal.ChronoUnit;

import cashpiles.util.Comparisons;

public class DateRange {

	private final boolean hasExplicitPeriod;
	private final Period period;
	private final LocalDate startDate;

	public DateRange(LocalDate startDate, LocalDate endDate) {
		sortDates(startDate, endDate);
		this.startDate = startDate;
		period = Period.ofDays(countDays(startDate, endDate));
		hasExplicitPeriod = false;
	}

	public DateRange(LocalDate startDate, Period period) {
		this.startDate = startDate;
		this.period = period;
		sortDates(startDate, endDate());
		hasExplicitPeriod = true;
	}

	public boolean contains(LocalDate date) {
		return startDate.compareTo(date) <= 0 && endDate().compareTo(date) >= 0;
	}

	public final LocalDate endDate() {
		return startDate.plus(period).minusDays(1);
	}

	@Override
	public boolean equals(Object other) {
		if (other == null) {
			return false;
		}
		if (!getClass().equals(other.getClass())) {
			return false;
		}
		return ((DateRange) other).startDate.equals(startDate) && ((DateRange) other).period.equals(period);
	}

	public boolean hasExplicitPeriod() {
		return hasExplicitPeriod;
	}

	@Override
	public int hashCode() {
		return startDate.hashCode() ^ period.hashCode();
	}

	public DateRange intersect(DateRange other) {
		var newStart = Comparisons.max(startDate, other.startDate);
		var newEnd = Comparisons.min(endDate(), other.endDate());
		sortDates(newStart, newEnd);
		return new DateRange(newStart, newEnd);
	}

	public DateRange next() {
		return new DateRange(startDate.plus(period), period);
	}

	public int numberOfDays() {
		return countDays(startDate, endDate());
	}

	public boolean overlaps(DateRange other) {
		var newStart = Comparisons.max(startDate, other.startDate);
		var newEnd = Comparisons.min(endDate(), other.endDate());
		return newStart.compareTo(newEnd) <= 0;
	}

	public Period period() {
		return period;
	}

	public LocalDate startDate() {
		return startDate;
	}

	public String toString() {
		return startDate.toString() + "+" + period + "->" + endDate().toString();
	}

	private static int countDays(LocalDate startDate, LocalDate endDate) {
		var days = ChronoUnit.DAYS.between(startDate, endDate) + 1;
		return BigInteger.valueOf(days).intValueExact();
	}

	private static void sortDates(LocalDate startDate, LocalDate endDate) {
		if (endDate.compareTo(startDate) < 0) {
			throw new DateTimeException("End date is before start date");
		}
	}

}
